aten/src/ATen/core/boxing/KernelFunction.h

```diff
@@ -0,0 +1,466 @@
+#pragma once
+
+#include <ATen/core/stack.h>
+#include <c10/util/TypeList.h>
+#include <ATen/core/boxing/kernel_functor.h>
+#include <ATen/core/boxing/kernel_function.h>
+#include <ATen/core/boxing/kernel_lambda.h>
+
+namespace c10 {
+
+namespace detail {
+template<class Type>
+constexpr uint64_t hashType() {
+  return typeid(Type).hash_code()
+      + 1000 * std::is_lvalue_reference<Type>::value
+      + 5000 * std::is_rvalue_reference<Type>::value
+      + 10000 * std::is_const<guts::remove_reference_t<Type>>::value
+      + 15000 * std::is_volatile<guts::remove_reference_t<Type>>::value
+      ;
+}
+template<class TypeList> struct hashTypeList_ final {};
+template<class Head, class... Tail>
+struct hashTypeList_<guts::typelist::typelist<Head, Tail...>> final {
+  static constexpr uint64_t call(uint64_t index) {
+    return 1000000 * index * hashType<Head>() + hashTypeList_<guts::typelist::typelist<Tail...>>::call(index + 1);
+  }
+};
+template<>
+struct hashTypeList_<guts::typelist::typelist<>> final {
+  static constexpr uint64_t call(uint64_t index) {
+    return 0;
+  }
+};
+
+template<class TypeList>
+constexpr uint64_t hashTypeList() {
+  return hashTypeList_<TypeList>::call(1);
+}
+
+// Take a function signature and produce a hash value depending on its
+// argument and return types. For the same function signature and while
+// running the same executable, this will always produce the same hash.
+// A different compiler/OS might generate different hashes, so don't use
+// these for serialization, but they're good for error checking the casting
+// of void* function pointers into actual typed function pointers.
+// Note that there it is not perfect error checking, two different signatures
+// might have the same hash, but it is probably good enough.
+template<class FuncSignature>
+constexpr uint64_t hashFunctionSignature() {
```

Comment:
Can't you just build a completely precise struct describing the function signature in question? Rather than doing an approximate hash...

Reply:
I could but that would be expensive to compare at runtime. Each time you call KernelFunction::callUnboxed, it's going to compare the args you call it with with the args of the functor that was used to create the KernelFunction. Doing that arg-by-arg is expensive, but computing a 64bit hash at compile time and at runtime just comparing two 64bit values is cheap.